logstash-core/lib/logstash/instrument/metric.rb

```diff
@@ -8,87 +8,102 @@ class MetricNoKeyProvided < MetricException; end
+      collector.push(namespace, key, :counter, :increment, value)
     end
 
-    def decrement(key, value = 1)
+    def decrement(namespace, key, value = 1)
       validate_key!(key)
-      collector.push(namespace_information, key, :counter, :decrement, value)
+      collector.push(namespace, key, :counter, :decrement, value)
     end
 
-    # might be worth to create a block interface for time based gauge
-    def gauge(key, value)
+    def gauge(namespace, key, value)
       validate_key!(key)
-      collector.push(namespace_information, key, :gauge, :set, value)
+      collector.push(namespace, key, :gauge, :set, value)
     end
 
-    def time(key)
+    def time(namespace, key)
       validate_key!(key)
 
       if block_given?
-        timer = TimedExecution.new(self, key)
+        timer = TimedExecution.new(self, namespace, key)
         content = yield
         timer.stop
         return content
       else
-        TimedExecution.new(self, key)
+        TimedExecution.new(self, namespace, key)
       end
     end
 
-    def report_time(key, duration)
-      collector.push(namespace_information, key, :mean, :increment, duration)
+    def report_time(namespace, key, duration)
+      collector.push(namespace, key, :mean, :increment, duration)
     end
 
-    def namespace(sub_namespace)
-      raise MetricNoNamespaceProvided if sub_namespace.nil? || sub_namespace.empty?
-
-      new_namespace = namespace_information.clone
-      new_namespace << sub_namespace
-
-      Metric.new(collector, new_namespace)
+    # This method return a metric instance tied to a specific namespace
+    # so instead of specifying the namespace on every call.
+    #
+    # Example:
+    #   metric.increment(:namespace, :mykey, 200)
+    #   metric.increment(:namespace, :mykey_2, 200)
+    #
+    #   namespaced_metric = metric.namespace(:namespace)
+    #   namespaced_metric.increment(:mykey, 200)
+    #   namespaced_metric.increment(:mykey_2, 200)
+    # ```
+    #
+    # @param name [Array<String>] Name of the namespace
+    # @param name [String] Name of the namespace
+    def namespace(name)
+      raise MetricNoNamespaceProvided if name.nil? || name.empty?
+
+      NamespacedMetric.new(self, name)
     end
 
-    def self.create(namespace, collector = LogStash::Instrument::Collector.instance)
-      Metric.new(collector, namespace)
+    # Create a Metric instrance using the default Collector singleton reference
+    #
+    #
+    def self.create(collector = LogStash::Instrument::Collector.instance)
+      Metric.new(collector)
     end
 
     private
     def validate_key!(key)
       raise MetricNoKeyProvided if key.nil? || key.empty?
     end
+
+    # Allow to calculate the execution of a block of code.
+    # This class support 2 differents syntax a block or the return of
+    # the object itself, but in the later case the metric wont be recorded
+    # Until we call `#stop`.
+    #
+    # @see LogStash::Instrument::Metric#time
+    class TimedExecution
```

Comment:
I understand this class is under private scope, right? I wonder what happen in the case of the time function returning an instance of ```TimedExecution``` would this lead to some problem?